include/openssl/e_os2.h

```diff
@@ -165,7 +165,7 @@ extern "C" {
  * have some generally sensible values.
  */
 
-# if defined(OPENSSL_SYS_WINDOWS) && defined(OPENSSL_OPT_WINDLL)
+# if defined(OPENSSL_SYS_WINDOWS) /* goes wrong: && defined(OPENSSL_OPT_WINDLL) */
```

Comment:
Would you mind elaborating?  Mind you, these mechanisms are ooooooold, and my memory is failing me.  But the OPENSSL_OPT_WINDLL stuff was set in place through commits 5cac59b6c1238ffc911c054b51db42c3aca3aaa2 and 595cc5b4b86327fba7bb0570294f7fb05987485b, would you say this is invalid today?

Reply:
I suspect that this dependency on `OPENSSL_OPT_WINDLL` is somehow outdated; at least when I built with 'shared' (DLLs) with VS2014 the  `__declspec(` variants did not fire, and removing the condition fixed it.

Reply:
This is public header and it's exactly place for arbitrary tweaking. One can argue that these two macros don't belong in this header at all(*), and move it to private header, but it's not something that would be appropriate to do in minor release.

(*) I mean application programmer has no control [or say] over symbol visibility in library, so why would it have to be in public header?

Reply:
> removing the condition fixed it.

Fixed what specifically? I mean what was it that didn't work?

Reply:
When experimenting with moving function definitions from `apps/` to the tentative new library in `secapp/` as motivated in #4992, I faced trouble with the external variable `bio_err` which is defined in `apps/openssl.c`. My first solution was to move this variable to 'secapp/' as well and to make it visible to any users of the new library, in particular code in `apps/`, via an `extern` declaration in the new `include/openssl/apps.h`. This should work straightforwardly, but for Windows native builds the linker reported the symbol as not defined. It took me a while to find out what the reason was: the two lines

        #  define OPENSSL_EXPORT extern __declspec(dllexport)
        #  define OPENSSL_EXTERN extern __declspec(dllimport)

were not active. After disabling the condition `defined(OPENSSL_OPT_WINDLL)` the linker was happy.
I wonder why this problem has not popped up before - maybe because to far OpenSSL libs do not provide variables but just functions and the Visual Studio linker is smarter handling functions? 

(In the end, I chose a different, more flexible solution: the `secapp` library has its own copy of the `bio_err` variable which may be manipulated by users of the library via a set/get function.
Therefore the fix regarding `defined(OPENSSL_OPT_WINDLL)` is not strictly needed for the library.)

In any case, for whatever reason, for my VS2014 setup `_WINDLL` is undefined although I used `Configure VC-WIN32 shared`. Would a better fix be to add `-D_WINDLL` to the `CFLAGS` of such a configuration? Then `e_os2.h` could stay as it was.

Reply:
As already mentioned here and in #4992, there are things that are not exactly appropriate to do in minor releases. So I'd suggest (insist really) to revert this modification. One should also recognize that there are things in public headers that shouldn't really be there, unjustified at times, yet kept for compatibility reasons. And key things about them is not to dig hole deeper, by fixing them up, but to abstain and explore other, more long-term options.

Reply:
As for _WINDLL. Its usage appears rather as misconception of some kind... It's kind of implied that it's a predefine [in MSVC] when it's not or even been. This kind of qualifies it for "don't-dig-deeper"...

Reply:
I always thought the intention was that those who wanted that particular declaration style should configure with an added `-D_WINDLL`...  I'd like to point out that we have this as part of the mingw and cygwin config targets

Reply:
Well, it gets handy when you link engine dlls. In MSVC case it's handled by inline .def file that exports bind_engin and v_check, while migw relies on declspec(dllexport). It's kind of opportunistic approach, and one can argue that it could be replaced with something unified, or at least something that is not something like _WINDLL that actually has some meaning to MFC. In either case, it doesn't actually belong in public header [as it's something that is meaningful only during build time]. so suggestion still is leave it alone and make mental note till 1.2.0 time. BTW, latter doesn't mean that one can't introduce "future-proof" constructs in 1.1.1, as long as they remain private, right?

Reply:
I agree with @dot-asm.  Please revert this mod for this PR.  Not denying the possible quality, but this isn't the right place for it.

Reply:
All right, reverted this. 

Yet how then solve the issue that at least for some Visual Studio instances `_WINDLL` is not automatically defined in MSVC when building DLLs? I am not an expert on Microsoft tools (I'm just forced using them in addition to Unix-like ones), but as far as I understand for instance https://msdn.microsoft.com/en-us/library/aa235516(v=vs.60).aspx, whenever the `/MD` switch or a similar one is present this should imply that `_WINDLL` is defined automatically, while at least for my instance of VS2014 it is not. So why not pragmatically, as a just-in-case workaround, add

        shared_cflag     => add("-D_WINDLL"),

to the `"VC-common"´ section of `Configurations/10-main.conf`?

In case you agree on this, in which branch should I put it?
Else, where to track such issues such that they don't get lost and will get dealt with in a more favorable moment?

Reply:
As for msdn reference. If you read it carefully, you won't actually find anywhere that `_WINDLL` is a *pre-defined* macro. It's rather something that is or has to be added to MFC[?] project. Though there is pre-defined macro that is associated with /MD, but it's not `_WINDLL`, it's `_DLL`. Keep in mind that it's MSVC-specific, while MSVC is not the only compiler we support.

As for what to do. Right thing is to do something that is a) not [potentially] confusing, b) works with either compiler. `_WINDLL` arguably doesn't meet a) criterion. And the fact that it's used with mingw doesn't mean that it *has to* be used for all eternity. Next question is what is it that needs to be solved... But that would have to something that would have to discussed next year :-)

Reply:
Thank you for your comments on my above questions.